File: prism/modules/learning_log.py
from __future__ import annotations
import time
import logging
from typing import Optional, TypedDict
from prism.config import PrismConfig
from prism.engines.grounding import GroundingAccumulator

log = logging.getLogger(__name__)


class CIAERRecord(TypedDict, total=False):
    timestamp: float
    cause: str
    intuition: str
    confidence_in_voice: str   # SKILL | RULE | KNOWLEDGE
    action: str
    effect: str
    reaction: str
    salience: float
    result: str
    shadow_actions: list
    model_revision: str        # optional; precious when present
    # Pre-ENV fields
    pre_env_time_of_day: float
    pre_env_companion_mood: str
    pre_env_child_recent_concepts: list


class LearningLog:
    """
    Writes CIAER+ records immediately (append-only).
    OGC HARD RULE: grounding_accumulator.record_reappearance() is only called
    when learning evidence exists — never on engagement-only events.
    """

    def __init__(self, config: PrismConfig,
                 grounding: GroundingAccumulator) -> None:
        self._cfg = config
        self._grounding = grounding
        self._records: list[CIAERRecord] = []
        self._store = None   # set by persistence layer after init

    def set_store(self, store) -> None:
        self._store = store

    def record_event(self, record: CIAERRecord) -> None:
        record['timestamp'] = record.get('timestamp', time.time())
        self._records.append(record)

        if self._store:
            try:
                self._store.append(record)
            except Exception as e:
                log.error("Failed to persist CIAER record: %s", e)

        # OGC enforcement: only update grounding on learning evidence
        concept = record.get('cause', '')
        has_model_revision = bool(record.get('model_revision'))
        result = record.get('result', '')
        new_context = record.get('action', '')

        # First exposure
        if concept:
            self._grounding.record_exposure(concept)

        # Reappearance with learning evidence: model revision OR non-trivial result
        has_learning_evidence = has_model_revision or (
            result and result not in ('', 'none', 'no_change')
        )
        if has_learning_evidence and concept:
            self._grounding.record_reappearance(
                concept=concept,
                new_context=new_context,
                has_model_revision=has_model_revision,
            )

    def get_recent(self, n: int = 50) -> list[CIAERRecord]:
        return self._records[-n:]

    def get_by_concept(self, concept: str) -> list[CIAERRecord]:
        return [r for r in self._records if r.get('cause') == concept]

    def get_all(self) -> list[CIAERRecord]:
        return list(self._records)

File: android/engine/src/main/kotlin/com/cappsconsulting/prism/engine/learninglog/LearningLog.kt
package com.cappsconsulting.prism.engine.learninglog

import com.cappsconsulting.prism.engine.grounding.GroundingAccumulator
import kotlinx.serialization.Serializable

/**
 * Direct port of `prism/modules/learning_log.py::CIAERRecord` (a `TypedDict` in the
 * original; a sealed, serializable data class here — Kotlin has no structural typing,
 * and this record crosses the sync boundary, Doc 3.0 §3.1, so it needs a stable wire
 * shape with `kotlinx.serialization` rather than an open bag of `Any?`).
 *
 * This is CIAER+ written twice over (Genotype "Schema & memory commitments"): once
 * for the child as the subject of `cause`/`intuition`/`effect`, and a second time in
 * the `pre_env_*` fields, which capture the *companion's* own pre-environment state —
 * the second CIAER agent the schema was built to support.
 */
@Serializable
data class CIAERRecord(
    val timestamp: Double,
    val cause: String = "",
    val intuition: String = "",
    val confidenceInVoice: String = "",
    val action: String = "",
    val effect: String = "",
    val reaction: String = "",
    val salience: Double = 0.0,
    val result: String = "",
    val shadowActions: List<String> = emptyList(),
    /** Optional; precious when present — strong learning-evidence signal (see [LearningLog.recordEvent]). */
    val modelRevision: String? = null,
    val preEnvTimeOfDay: Double = 0.0,
    val preEnvCompanionMood: String = "",
    val preEnvChildRecentConcepts: List<String> = emptyList(),
)

/**
 * Direct port of `prism/modules/learning_log.py::LearningLog`.
 *
 * Writes CIAER+ records immediately, append-only — and is the single call site that
 * enforces the OGC hard rule: [GroundingAccumulator.recordReappearance] fires *only*
 * when [recordEvent] sees actual learning evidence (a model revision, or a non-trivial
 * result), never on engagement alone. Centralizing the rule here — rather than letting
 * every call site decide for itself whether to bump grounding — is what makes "optimize
 * the teaching, never the engagement" (Hard Line 10) structural rather than a
 * convention someone could forget at one of many call sites.
 *
 * @property onPersist Called for every record (replaces the Python `_store` /
 * `set_store` seam — the persistence layer is injected by whichever app wires this up;
 * see `:companion-app` `data/LearningLogStore`). Exceptions are caught and logged by
 * the caller's wiring, exactly as the Python wrapped `self._store.append` in try/except —
 * a persistence failure must never interrupt the live session.
 */
class LearningLog(
    private val grounding: GroundingAccumulator,
    private val onPersist: ((CIAERRecord) -> Unit)? = null,
) {
    private val records: MutableList<CIAERRecord> = mutableListOf()

    fun recordEvent(record: CIAERRecord) {
        records.add(record)
        onPersist?.invoke(record)

        // OGC enforcement — the load-bearing logic of this whole class.
        val concept = record.cause
        val hasModelRevision = !record.modelRevision.isNullOrEmpty()
        val result = record.result
        val newContext = record.action

        if (concept.isNotEmpty()) {
            grounding.recordExposure(concept)
        }

        val hasLearningEvidence = hasModelRevision ||
            (result.isNotEmpty() && result != "none" && result != "no_change")

        if (hasLearningEvidence && concept.isNotEmpty()) {
            grounding.recordReappearance(
                concept = concept,
                newContext = newContext,
                hasModelRevision = hasModelRevision,
            )
        }
    }

    fun getRecent(n: Int = 50): List<CIAERRecord> = records.takeLast(n)
    fun getByConcept(concept: String): List<CIAERRecord> = records.filter { it.cause == concept }
    fun getAll(): List<CIAERRecord> = records.toList()
}
